File: PhaseGate/StudentGrade/StudentGrade.java
import java.util.Scanner;
public class StudentGrade{

	static Scanner input = new Scanner(System.in);



	
	public static void main(String... args){

		 collectStudentsInputs();

		
	}
	public static void collectStudentsInputs(){

		System.out.println("Enter number of Students : ");
		int studentNumber = input.nextInt();
		System.out.println("Enter number of Subjects : ");
		int numberOfSubjects = input.nextInt();

		String[] studentNames = new String[studentNumber];
		String[] sortstudentNames = new String[studentNumber];
		String[] studentSubjects = new String[numberOfSubjects];
		int[][] studentGrade = new int[studentNumber][numberOfSubjects]; 
		int[][] passesAndfaliures = new int[studentNumber][2];

		input.nextLine();
		for(int count = 0; count<studentNames.length ; count++){
			
			System.out.println("Enter names of Students : ");
			studentNames[count] = input.nextLine();
		}	
		
		for(int count = 0; count<studentSubjects.length ; count++){
		
			System.out.println("Enter names of subject Students : ");
			studentSubjects[count] = input.nextLine();
		}
		
		for(int count = 0; count<studentNumber ; count++){
				
			System.out.println("Enter grade "+ studentNames[count] );
			int failures = 0;
			int passes = 0;
			for(int counter = 0; counter<numberOfSubjects ; counter++){
				
				
				System.out.printf("Enter %s  grade :",studentSubjects[counter]);
				
					studentGrade[count][counter]  = input.nextInt();
					
					if (studentGrade[count][counter]<0 || studentGrade[count][counter]>100){
						System.out.print(" invalid !!! \nEnter a valid grade !!!");
						--counter;
					}
					if (studentGrade[count][counter] >= 45){
						passes +=1;
						passesAndfaliures[count][0] = passes;
					}
					if ( studentGrade[count][counter] < 45  ){
						failures+=1;
						passesAndfaliures[count][1] = failures ;
					}
			
				
				System.out.println("Saving>>>>>>>>>>>>>>>>>>>>");
			}
			

			
			
			
		}
		
	for(int count = 0 ; count < studentNumber ; count++){

		System.out.println(getSort(eachStudentTotal(studentGrade, studentNumber))[count]+" ******");

	}
	
	
	System.out.println("========================================================================== ");
	System.out.print("Student");
		for(int count=0 ; count<studentSubjects.length; count++){
			System.out.printf("	%s",studentSubjects[count]);
		
		}
	System.out.printf("%9s ","ToT");
	System.out.printf("%11s","AVE");
	System.out.printf("%14s ","pos");
	System.out.println();
	System.out.println("========================================================================== ");


		for(int count=0 ; count< studentNumber ; count++){
		
		System.out.print(studentNames[count]);
			for(int index = 0 ; index < studentSubjects.length ; index++){
				System.out.printf("	" + studentGrade[count][index] ); 
			} 
		System.out.printf("%10d	",eachStudentTotal(studentGrade, studentNumber)[count]);
		System.out.printf("%11.2f",((float)(eachStudentTotal(studentGrade, studentNumber)[count])/numberOfSubjects));
		System.out.printf("%14d",position(getSort(eachStudentTotal(studentGrade, studentNumber)) ,eachStudentTotal(studentGrade, studentNumber) )[count]+1);
		
		System.out.println();		
		}
		
		
		System.out.println("========================================================================== \n\n========================================================================== ");
		System.out.println();





		highestAndLowestPreSubject( numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade));
		totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade));
		passesAndFailuresOfEachSubjects( numberOfSubjects , studentNumber , eachSubjects(numberOfSubjects, studentNumber ,studentGrade));
		highestInteger(totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade)));
		lowestInteger(totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade)));
		
		




	
		System.out.println("STUDENT SUMMARY");
		System.out.println("--------------------------------------------------------------------------------");
		for(int index = 0; index < numberOfSubjects ; index++){
			
	
				System.out.println( studentSubjects[index] );
				System.out.print("Highest Scoring Student is ");
				for(int count=0; count < numberOfSubjects ; count++){
					if(highestAndLowestPreSubject(numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count][0] == studentGrade[count][index]){
						System.out.printf("%s%s Scoring :%d \n", studentNames[count], studentNames[count] , highestAndLowestPreSubject(numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count][0]);
					}
				} 
				System.out.print("lowest Scoring Student is ");
				for(int count=0; count < numberOfSubjects ; count++){

					if(highestAndLowestPreSubject( numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count][1] == studentGrade[count][index]){
						System.out.printf(" %s%s Scoring : %d \n ", studentNames[count] , studentNames[count] ,highestAndLowestPreSubject( numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count][1]);
					}
				}
				System.out.printf(" Total Score is : %d",totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[index]);
				System.out.println();
				System.out.printf(" Average Score Score is : %.3f%n",(float)totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[index]/studentNumber);
				System.out.printf(" Number of Passes: %d \n Number of failures is : %d%n",passesAndFailuresOfEachSubjects(numberOfSubjects , studentNumber , eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[index][0] ,passesAndFailuresOfEachSubjects(numberOfSubjects , studentNumber , eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[index][1] ); 
				System.out.println("--------------------------------------------------------------------------------");
				

		} 


		
		
		for(int count=0; count < numberOfSubjects ; count++){
			if (lowestInteger(totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade))) == totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count] ){
				System.out.printf("The hardest subject is %s with number %d\n ", studentSubjects[count], passesAndFailuresOfEachSubjects(numberOfSubjects , studentNumber , eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count][1] );
				
			}
		}
		for(int count=0; count < numberOfSubjects ; count++){
			if (highestInteger(totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade))) == totalPreSubject(numberOfSubjects,eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count] ){
				System.out.printf("The easiest subject is %s with number %d\n " , studentSubjects[count] , passesAndFailuresOfEachSubjects(numberOfSubjects , studentNumber , eachSubjects(numberOfSubjects, studentNumber ,studentGrade))[count][0] );
				
			}
		}
		
		for(int count = 0 ; count < numberOfSubjects ; count++){
			for(int index = 0 ; index < studentNumber ; index++){
				if (  overallHighest( highestAndLowestPreSubject(numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade)) , numberOfSubjects) == eachSubjects(numberOfSubjects, studentNumber ,studentGrade)[count][index]){
					System.out.printf("The ovarall highest score is Scored by %s in %s Scoring %d \n", studentNames[index], studentSubjects[count] , overallHighest( highestAndLowestPreSubject(numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade)) , numberOfSubjects)); 
				}
			}
		}


		for(int count = 0 ; count < numberOfSubjects ; count++){
			for(int index = 0 ; index < studentNumber ; index++){
				if ( overallLowest( highestAndLowestPreSubject(numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade)) , numberOfSubjects) == eachSubjects(numberOfSubjects, studentNumber ,studentGrade)[count][index]){
					System.out.printf("The ovarall lowest score is Scored by %s in %s Scoring %d \n", studentNames[index], studentSubjects[count] , overallLowest( highestAndLowestPreSubject(numberOfSubjects, eachSubjects(numberOfSubjects, studentNumber ,studentGrade)) , numberOfSubjects)); 
				}
			}
		}



		System.out.println("Class Summary");
		System.out.println("========================================================================");
		System.out.println("Best Graduating Student is : "); 
		System.out.print("========================================================================\n\n\n");
		System.out.println("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");		
		System.out.println("Worse Graduating Student is : "); 
		System.out.println("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
			
		System.out.println("========================================================================");
		System.out.println("Class Total : ");
		System.out.println("Class Average Score is : ");
		System.out.println("========================================================================");

		
	}


	public static int listTotal( int[] scoresList ){
		int total = 0;
		for(int scores = 0; scores<scoresList.length; scores++){
			total+=scoresList[scores]; 
		}
		return total;
	}
	
	public static int[] eachStudentTotal(int[][] studentGrade, int studentNumber){
		int[] studentTotal = new int[studentNumber];	
		for(int count = 0; count<studentTotal.length ; count++ ){
		
			studentTotal[count] = listTotal(studentGrade[count]); 
			
		}
		return studentTotal;
	}


	public static int[] getSort(int[] sortedStudentTotal){
		for(int count = 0; count < sortedStudentTotal.length-1 ; count++){ 
		
			for(int counter = 0; counter < sortedStudentTotal.length-1-count; counter++){

				if (sortedStudentTotal[counter] < sortedStudentTotal[counter + 1]){
				int container = sortedStudentTotal[counter + 1]; 
				sortedStudentTotal[counter + 1] = sortedStudentTotal[counter];
				sortedStudentTotal[counter] = container;
					
				
				}
			
			
			}
	
	
		} 
		return sortedStudentTotal;
	}

	public static int[] position(int[] sortedTotal , int[] studentTotal ){
		
		int[] position = new int[sortedTotal.length];

		for(int count = 0; count<sortedTotal.length ;count++){
				 
			for(int index = 0; index < sortedTotal.length; index++){
				
				if (sortedTotal[count] == studentTotal[index]){
					
					position[index] = count ;
					
				}
			}
		}
		return position;
	}
	
	

	public static int[][] eachSubjects(int numberOfSubjects, int studentNumber ,int[][] studentGrade){
		int[][] eachSubjects = new int[numberOfSubjects][studentNumber];
		for(int index = 0 ; index<numberOfSubjects ; index++){
			for(int count = 0 ; count<studentNumber ; count++){
				eachSubjects[index][count] = studentGrade[count][index];
			}		
		}
		return eachSubjects;
	}

	public static int[][] highestAndLowestPreSubject(int numberOfSubjects, int[][] eachSubjects){
		int[][] highestAndLowestPreSubject = new int [numberOfSubjects][2]; 
		for(int index=0; index<numberOfSubjects ; index++){
			
		highestAndLowestPreSubject[index][0] = highestInteger(eachSubjects[index]);
		highestAndLowestPreSubject[index][1] = lowestInteger(eachSubjects[index]);
		
		}
		return highestAndLowestPreSubject;
	}

	public static int highestInteger(int[] numbers){
		int largest = numbers[0];
		for(int count = 0; count<numbers.length;count++){
			if(numbers[count]>largest){
				largest = numbers[count];
			}
		} 
	return largest; 
	
	}

	public static int lowestInteger(int[] numbers){
		int lowest = numbers[0];
		for(int count = 0; count<numbers.length;count++){
			if(numbers[count]<lowest){
				lowest = numbers[count];
			}
		} 
	return lowest;
	
	}


	public static int[] totalPreSubject(int numberOfSubjects, int[][] eachSubjects){
				
		int[] totalPreSubject = new int [numberOfSubjects]; 
		for(int index=0; index<numberOfSubjects ; index++){
			totalPreSubject[index] = listTotal(eachSubjects[index]);
			
		}
	return totalPreSubject;
	}

	
	public static int[][] passesAndFailuresOfEachSubjects(int numberOfSubjects , int studentNumber, int[][] eachSubjects ){
		
		int[][] passesAndFailuresOfEachSubjects = new int[numberOfSubjects][2];
		for(int index=0;index<numberOfSubjects; index++){
			for(int innerIndex=0; innerIndex < studentNumber ;innerIndex++){
				if(eachSubjects[index][innerIndex] > 45){
					passesAndFailuresOfEachSubjects[index][0] +=1;
				}
				else{
					passesAndFailuresOfEachSubjects[index][1] +=1; 
				}
			}
		}
		return passesAndFailuresOfEachSubjects;	
	}
	
	
	public static int overallHighest(int[][] highestAndLowestPreSubject , int numberOfSubjects){
		int[] allHighest = new int[numberOfSubjects];
		for(int count = 0 ; count< numberOfSubjects ; count++){
			allHighest[count] = highestAndLowestPreSubject[count][0];
		}
		return highestInteger(allHighest);
			
	}
	
	public static int overallLowest(int[][] highestAndLowestPreSubject , int numberOfSubjects){
		int[] allLowest = new int[numberOfSubjects];
		for(int count = 0 ; count< numberOfSubjects ; count++){
			allLowest[count] = highestAndLowestPreSubject[count][1];
		}
		return lowestInteger(allLowest);
			
	}




	

}

File: PhaseGate/StudentGrade/studentgrade.py
studentGrades = {}
studentnumber = int(input("Enter number of Students :"))
numberOfSubjects = int(input("Enter number of Subjects :"))
studentNames = []
studentSubject = []


def get_Hardest_add_easiest_subject(studentgrades:dict,subjects:list):
	totalpersubjects = []
	subjectkey = []
	passes = []
	faliures = []
	for subject in subjects:
		subjectname = ''
		total = 0
		pas = 0
		fails = 0
		for student in studentGrades:
			if studentGrades[student][subject] >= 45:
				pas += 1
			else:
				fails += 1 
			total += studentGrades[student][subject]
			subjectkey = subject		
		totalpersubjects.append(total)
		subjectkey.append(subjectname)
		passes.append(pas)
		faliures.append(fails)

	highest = get_highest(totalpersubjects)
	lowest = get_lowest(totalpersubjects)
	hardest_subject = " "
	hardest_failure = 0
	easiest_passes = 0
	for count in range len(subjectkey):
		if get_highest(totalpersubjects) == totalpersubjects[count]:
			easiest_subject = subjectkey[count]
			easiest_passes = passes[count]
		if get_lowest(totalpersubjects) == totalpersubjects[count]:
			hardest_subject = subjectkey[count]
			hardest_failure = failures

	return highest, lowest, easiest_subject , hardest_subject 

def get_lowest(numbers:list):
	lowest = 100
	for count in numbers:
		if count < lowest:
			lowest = count
	return lowest


def get_highest(numbers:list):
	highest = 0
	for count in numbers:
		if count > highest:
			highest = count 
	return highest 
	

def get_list_total(numbers:list):
	total = 0 
	for number in numbers:
		total += number
	return total

def overall_lowest_score(studentGrades:dict ):
	lowest = 100
	display_subject = " "
	name = " "
	for student in studentGrades:
		for subject in studentSubject:
			if studentGrades[student][subject] < lowest:
				lowest = studentGrades[student][subject]
				display_subject = subject
				name = student
			
	return lowest,	display_subject, name



def overall_highest_score(studentGrades:dict ):
	highest = 0
	display_subject = " "
	name = " "
	for student in studentGrades:
		for subject in studentSubject:
			if studentGrades[student][subject] > highest:
				highest = studentGrades[student][subject]
				display_subject = subject
				name = student
			
	return highest,	display_subject, name
	

	
def subjectTotal(studentGrades:dict):
	totals = []
	for subject in studentGrades:
		total = 0
		for student in studentGrades[subject]:
			total += studentGrades[subject][student]	
		totals.append(total)
	return totals

def passesAndFailures(studenGrades:dict, subjects:list):
	passes = []
	fails = []
	for subject in subjects:
		passed = 0
		failed = 0
		for student in studentGrades:
			if studentGrades[student][subject] >= 45 :
				passed += 1
			else:
				failed += 1
		passes.append(passed)
		fails.append(failed)
	return passes , fails

def getLowestPerSubject(studentgrades:dict,subjects:list):
	lowestStudents = []
	subjectkey = []
	for subject in subjects:
		lowest = 100
		subjectname = ''
		for student in studentGrades:
			if studentGrades[student][subject] < lowest :
				lowest = studentGrades[student][subject]
				subjectname = student
		lowestStudents.append(lowest)
		subjectkey.append(subjectname)
		
				
	return lowestStudents, subjectkey



def getHighestPerSubject(studentgrades:dict,subjects:list):
	highestStudents = []
	subjectkey = []
	for subject in subjects:
		highest = 0
		subjectname = ''
		for student in studentGrades:
			if studentGrades[student][subject] > highest :
				highest = studentGrades[student][subject]
				subjectname = student
		highestStudents.append(highest)
		subjectkey.append(subjectname)
		
				
	return highestStudents, subjectkey



def getPosition(numbers:list):
	position = numbers
	sortedNumbers = sorted(numbers) 
	sortedNumbers.sort(reverse=True)
	count = 1
	for grade in sortedNumbers:
		index = 0
		for check in range(len(numbers)):

			if grade == numbers[index]:
				position[index] = count
			index += 1
		count += 1
	return position


def getEachStudentTotal(studentSubject:list , studentGrades:dict):
	total = []
	for student in studentGrades:
		sum = 0
		for grade in studentSubject :
			sum += studentGrades[student][grade]
		total.append(sum)
			
	return list(total)


for count in range(studentnumber):
	name = input(f"Enter {count+1} Student names : ")
	studentNames.append(name)

for count in range(numberOfSubjects):
	subject = input(f"Enter Students Subjects {count+1} : ") 
	studentSubject.append(subject)

for name in studentNames:
	studentGrades[name] = {}
	for subject in studentSubject:
		grade = int(input(f"Enter {name} grade for {subject} : "))
		#if grade <= 100 and grade >= 0 :
		studentGrades[name][subject] = grade 
		print("Saving >>>>>>>>>>>>>>>>>>>>>>>>>>")
		print("Saved Suceessfully")
		#else :
		print("invalid input !!!")


print("=================================================================================")
print("Student",end="")
for subject in studentSubject:
	print(f"	{subject}",end="" )
print("\t",end="")

print("%-10s%-11s%-13s " %("Total" , "AVE" ,"POS"))
print("=================================================================================")
count = 0
for subjects in studentGrades :
	print(f"{subjects}",end="")
	for grade in studentSubject :
		print("\t%s" %(studentGrades[subjects][grade]),end="")
	print("%10d"%(getEachStudentTotal(studentSubject,studentGrades)[count]),end="")
	print("%11d"%((getEachStudentTotal(studentSubject,studentGrades)[count])/(numberOfSubjects)),end="")
	print("%13d"%(getPosition(getEachStudentTotal(studentSubject,studentGrades))[count]))
	count += 1
	print()

print()
print("Student Summary")
print()
highestScores , highestStudents = getHighestPerSubject(studentGrades , studentSubject)
lowestScores , lowestStudents = getLowestPerSubject(studentGrades , studentSubject)
numberOfPasse , numberOfFailures = passesAndFailures(studentGrades, studentSubject)
totals = subjectTotal(studentGrades)
overall_highest, overall_highest_subject, overall_highest_name = overall_highest_score(studentGrades)  
overall_lowest, overall_lowest_subject, overall_lowest_name = overall_lowest_score(studentGrades)
overall_total= get_list_total(getEachStudentTotal(studentSubject, studentGrades))


highest_subject, lowest_subject, easiest_subject , hardest_subject  = get_Hardest_add_easiest_subject(studentgrades,subjects)




count = 0
for subject in studentSubject:
	print()
	print(f"{subject}")
	print(f"the highest scoring is  {highestStudents[count]} scoring {highestScores[count]} ")
	print(f"the lowest scoring student is {lowestStudents[count]} {lowestScores[count]} scoring ")
	print(f"total is : {totals[count]}")
	print(f"Average is : {round(totals[count]/studentnumber , 3)}")
	print(f"number of passes : {numberOfPasse[count]}")
	print(f"number of failes : {numberOfFailures[count]}")
	print()
	print("--------------------------------------------------------------------------------------------")
	count += 1


get_highest(numberOfPasse)
get_lowest(numberOfPasse)
#print(f"the hardest Subject is {} with {} faliures")
#print(f"The easiest Subject is {} with {} passes")
print(f"the overall Highest score scored by {overall_highest_name} in {overall_highest_subject} scoring {overall_highest}")
print(f"the overall lowest score scored by { overall_lowest_name} in {overall_lowest_subject} scoring {overall_lowest}")
print("==========================================================================================")
print()

print("Class Summary")
print("==========================================================================================")
#print("best Graduating Student is : {} Scoring {} ")
print("==========================================================================================")
print()
print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
#print("Worse Graduationg Student is : {} scpring {}")
print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
print()
print("==========================================================================================")
print(f"Class Total Score is : {overall_total} ")
print(f"Class Aversge Score is : {overall_total/studentnumber} ")
print("==========================================================================================")
